File: GGAndroid_2.0/GGCore/GGUtils/src/main/java/io/ganguo/utils/util/Screens.java
package io.ganguo.utils.util;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.Surface;
import android.view.WindowManager;

import io.ganguo.utils.bean.Globals;

/**
 * <p>
 * 屏幕操作工具类
 * </p>
 * Created by leo on 2018/8/6.
 */

public class Screens {
    public static final String LOGGER_TAG = Screens.class.getSimpleName();

    private Screens() {
        throw new Error(Globals.ERROR_MSG_UTILS_CONSTRUCTOR);
    }

    /**
     * Screen DisplayMetrics
     *
     * @param context
     * @return
     */
    public static DisplayMetrics getDisplayMetrics(Context context) {
        if (context == null) {
            return null;
        }
        return context.getResources().getDisplayMetrics();
    }


    /**
     * Screen Display
     *
     * @param context
     * @return
     */
    public static Display getDisplay(Context context) {
        if (context == null) {
            return null;
        }
        WindowManager manager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = manager.getDefaultDisplay();
        return display;
    }


    /**
     * 获取字体ScaledDensity值
     *
     * @param context
     * @return
     */
    public static float getScaledDensity(Context context) {
        DisplayMetrics displayMetrics = getDisplayMetrics(context);
        return displayMetrics.scaledDensity;
    }

    /**
     * 获取屏幕高度
     *
     * @param context
     * @return
     */
    public static int getScreenHeight(Context context) {
        DisplayMetrics displayMetrics = getDisplayMetrics(context);
        return displayMetrics.heightPixels;
    }

    /**
     * 获取屏幕宽度
     *
     * @param context
     * @return
     */
    public static int getScreenWidth(Context context) {
        DisplayMetrics displayMetrics = getDisplayMetrics(context);
        return displayMetrics.widthPixels;
    }


    /**
     * 当前是否是横屏
     *
     * @param context
     * @return
     */
    public static final boolean isLandscape(Context context) {
        return context.getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
    }

    /**
     * 当前是否是竖屏
     *
     * @param context
     * @return
     */
    public static final boolean isPortrait(Context context) {
        return context.getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT;
    }


    /**
     * 获取当前窗口的旋转角度
     *
     * @param activity
     * @return
     */
    public static int getWindowDisplayRotation(Activity activity) {
        switch (activity.getWindowManager().getDefaultDisplay().getRotation()) {
            case Surface.ROTATION_0:
                return 0;
            case Surface.ROTATION_90:
                return 90;
            case Surface.ROTATION_180:
                return 180;
            case Surface.ROTATION_270:
                return 270;
            default:
                return 0;
        }
    }


    /**
     * 设置Activity为全屏
     *
     * @param activity
     */
    public static void setFullScreen(Activity activity) {
        WindowManager.LayoutParams params = activity.getWindow().getAttributes();
        params.flags |= WindowManager.LayoutParams.FLAG_FULLSCREEN;
        activity.getWindow().setAttributes(params);
        activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
    }
}

File: GGAndroid_4.0/core/support/support-screen/src/main/kotlin/io/ganguo/screen/Screens.kt
@file:JvmName("Screens")
@file:Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")

/**
 * <pre>
 *     author : leo
 *     time   : 2020/05/23
 *     desc   : Screen Utils
 * </pre>
 */
package io.ganguo.screen

import android.content.Context
import android.content.res.Configuration
import android.os.Build
import android.util.DisplayMetrics
import android.view.Display
import android.view.WindowManager


/**
 * Context is Null
 * @receiver Context?
 * @return Boolean
 */
fun Context?.isNull(): Boolean {
    return this == null
}

/**
 * Context is Not Null
 * @receiver Context
 * @return Boolean
 */
fun Context?.isNotNull(): Boolean {
    return this != null
}

/**
 * get DisplayMetrics from Context
 * @receiver Context?
 * @return DisplayMetrics?
 */
fun Context?.displayMetrics(): DisplayMetrics? {
    return this!!.resources?.displayMetrics
}


/**
 * get Display from Context
 * @receiver Context?
 * @return Display
 */
fun Context?.display(): Display? {
    return this?.windowService()?.defaultDisplay
}


/**
 * get WindowManager from Context
 * @receiver Context?
 * @return WindowManager
 */
fun Context?.windowService(): WindowManager? {
    return this?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
}

/**
 * get font size scaledDensity from Context
 * @receiver Context?
 * @return Float?
 */
fun Context?.scaledDensity(): Float {
    return displayMetrics()?.scaledDensity ?: 0f
}

/**
 * get screen width from Context
 * @receiver Context?
 * @return Int?
 */
fun Context?.screenWidth(): Int {
    return displayMetrics()?.widthPixels ?: 0
}

/**
 * get screen width from Context
 * @receiver Context?
 * @return Int?
 */
fun Context?.screenHeight(): Int {
    return displayMetrics()?.heightPixels ?: 0
}

/**
 * get Configuration from Context
 * @receiver Context?
 * @return Configuration?
 */
fun Context?.configuration(): Configuration? {
    return this?.resources?.configuration
}

/**
 * current Screen is landscape from Context
 * @receiver Context?
 * @return Boolean
 */
fun Context?.isLandscape(): Boolean {
    val configuration = configuration() ?: return false
    return configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
}

/**
 * current Screen is portrait from Context
 * @receiver Context?
 * @return Boolean
 */
fun Context?.isPortrait(): Boolean {
    return !isLandscape()
}


/**
 * get statusBar Height  from Context
 * @receiver Context?
 * @return Int
 */
fun Context?.statusBarHeight(): Int {
    val default = 38
    if (isNull()) {
        return default
    }
    var height = default
    try {
        val c = Class.forName("com.android.internal.R\$dimen")
        val obj = c.getDeclaredConstructor().newInstance()
        val field = c.getField("status_bar_height")
        val x: Int = field[obj].toString().toInt()
        height = this?.resources?.getDimensionPixelSize(x) ?: default
    } catch (e1: Exception) {
        e1.printStackTrace()
    }
    return height
}


/**
 *  is it versionCode version or above
 * @param versionCode Int
 * @return Boolean
 */
fun isAtLeastSdkVersion(versionCode: Int): Boolean {
    return Build.VERSION.SDK_INT >= versionCode
}

